dashboard/app/assets/javascripts/header.js.erb

```diff
@@ -2,110 +2,575 @@
+      if (parts) {
+        window.headerProgress = new HeaderProgress( $('.header-wrapper'), {
+          script_name: parts[1],
+          stage_id: parts[2],
+          level_id: parts[3]
+        });
+        return;
+      }
+
+      // Unknown route
+      return null;
+    }
+
+
+    return UIRouter;
+  })();
+
+
+
+
+  // class UIStore: this is a lightweight inheritable *ABSTRACT* class modeled after Flux
+  // UIStore is "offline-aware", meaning that it will load assets in a different way when
+  // it knows that there is no network connection.
+  UIStore = (function() {
+
+    function UIStore() {
+      this.save = __bind(this.save, this);
+      this.onAjaxFailure = __bind(this.onAjaxFailure, this);
+
+      this.value = null; // Empty store
+    }
+
+    // Get the value of this store
+    UIStore.prototype.value = function() { return this.value; }
+
+    // When this store changes, notify via callback(data)
+    // TODO: Maintain a list of callbacks
+    UIStore.prototype.notify = function(callback) {
+      this.notificationCallback = callback;
+    }
+
+    // Internal call to process any data that's being saved
+    UIStore.prototype.process = function(data) { return data; }
+
+    // Set the value of this store and notify the dependent
+    UIStore.prototype.save = function(data) {
+      this.value = this.process(data);
+
+      // Note: Flux uses emitters...
+      if (this.notificationCallback)
+        this.notificationCallback(this.value);
+    }
+
+    // Resolve a dynamic asset by combining the base asset filename + a serialized version of the args
+    // MUST override this.  It's okay to use args and dataType to determine the URL.
+    UIStore.prototype.resolveUrl = function(args, dataType) {
+      throw new Error("This store has not specified its URL.");
+    }
+
+    // Resolve a static asset by combining the base asset filename + a serialized version of the args
+    // MUST override this if you want to support offline access.
+    // Example: return "/data/file_" + args.id + "." + dataType;
+    UIStore.prototype.resolveFilename = function(args, dataType) {
+      throw new Error("This store does not support offline mode.");
+    }
+
+    // Resolve the AJAX parameters for a dynamic request
+    // Okay to override this to, for instance, move some of the args into the URL, etc.
+    UIStore.prototype.resolve = function(args, dataType) {
+      return {
+        url: this.resolveUrl(args, dataType),
+        data: args
+      };
+    }
+
+    // What are the AJAX parameters for this callback?
+    // If you override this, the override should be offline-aware.
+    UIStore.prototype.resolveAjax = function(args, dataType) {
+      var ajax;
+
+      if (!this.offline)
+        return this.resolve(args);
+
+      // Offline case
+      return {
+        url: this.resolveFilename(args, dataType)
+      };
+    }
+
+    UIStore.prototype.onAjaxFailure = function(xhr, status, message) {
+      // TODO
```

Comment:
TODO.  Do we plan on ever handling the AJAX failure?  If so, let's implement it.  If not, let's remove.